### PR TITLE
Adding initial events from fendermint

### PR DESCRIPTION
**Pull request by cryptoAtwill**
Adding initial set of events for logging from fendermint.

The approach is rather simple at this stage to avoid complex structs/macro. The event type is centralised in `core` crate then other crate can call `emit!` to emit the event in the format that is most suitable.

### CODE REVIEW

File: fendermint/vm/core/src/event.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+#[derive(strum_macros::Display)]
```
Comment (raulk) [inline fendermint/vm/core/src/event.rs:4]: I would move this to a top-level `fendermint/event` package, and rename it to `EventType`.

File: fendermint/vm/core/Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "fendermint_vm_core"
-description = "Core data structures shared between genesis and messages"
+description = "Core data structures shared between the vm"
```
Comment (raulk) [inline fendermint/vm/core/Cargo.toml:3]: ?

File: Cargo.toml [hunk 3]
```diff
@@ -200,6 +201,8 @@ cid = { version = "0.10.1", default-features = false, features = [
 
 frc42_dispatch = "6.0.0"
 
+strum_macros = "0.25.2"
```
Comment (raulk) [inline Cargo.toml:204]: You already have `strum` in the workspace, which exports the macros; no need to import this crate specifically.

File: fendermint/vm/core/src/event.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+pub enum VMEvent {
+    NewParentNullRound,
+    NewParentNonNullRound,
```
Comment (raulk) [inline fendermint/vm/core/src/event.rs:7]: These should be the same event with a `null_round` bool field.

File: fendermint/vm/genesis/Cargo.toml [hunk 5]
```diff
@@ -26,7 +26,7 @@ fendermint_actor_eam = { workspace = true }
 
 fendermint_crypto = { path = "../../crypto" }
 fendermint_testing = { path = "../../testing", optional = true }
-fendermint_vm_core = { path = "../core" }
+fendermint_vm_core = { workspace = true }
```
Comment (raulk) [inline fendermint/vm/genesis/Cargo.toml:29]: Why this change? What do we gain?

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs [hunk 6]
```diff
@@ -116,6 +121,14 @@ where
         power_diff(curr_power_table, next_power_table)
     };
 
+    emit!(
+        VMEvent::NewBottomUpCheckpoint,
+        height = height.value(),
+        block_hash = hex::encode(block_hash),
```
Comment (raulk) [inline fendermint/vm/interpreter/src/fvm/checkpoint.rs:127]: Perhaps we add a fmt::Display implementation to whatever type `block_hash` is instead? I see us doing this in several places.

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs [hunk 7]
```diff
@@ -116,6 +121,14 @@ where
         power_diff(curr_power_table, next_power_table)
     };
 
+    emit!(
+        VMEvent::NewBottomUpCheckpoint,
```
Comment (raulk) [inline fendermint/vm/interpreter/src/fvm/checkpoint.rs:125]: What level is this emitted at exactly? And what's the name of this field?

**Review by raulk — COMMENTED** [review]
I think this pattern works, but let's pay attention to:

1. What events end up looking like in practice (can you paste the output?).
2. DRY implementation, e.g. we might need to add some fmt::Display implementations for a few types.
3. Clean event schema, let's try not to explode on event types, and instead favour fields to indicate variants.

File: fendermint/vm/genesis/Cargo.toml (hunk 5 above)
Comment (cryptoAtwill) [reply fendermint/vm/genesis/Cargo.toml:29]: oh, it's just using the workspace build, so that there is just one build, so the crates all use the same version. Should not have any effect.

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs (hunk 6 above)
Comment (cryptoAtwill) [reply fendermint/vm/interpreter/src/fvm/checkpoint.rs:127]: block_hash is `Vec<u8>` or slice, unless we crate a wrapper type, it's kind of difficult to impl that.

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs (hunk 7 above)
Comment (cryptoAtwill) [reply fendermint/vm/interpreter/src/fvm/checkpoint.rs:125]: I used `info!` and field name is `event`.

**Comment by cryptoAtwill** [comment]
@fridrik01 The rest of the metrics are from harder to get, most are from combetbft I think. The ABCI ones we can use `#[instrument]`, should be easily doable.

**Review by fridrik01 — APPROVED** [review]
